presto-main/src/main/java/io/prestosql/sql/planner/ScopeAware.java

```diff
@@ -0,0 +1,156 @@
+        this.node = requireNonNull(node, "node is null");
+    }
+
+    public static <T extends Node> ScopeAware<T> scopeAwareKey(T node, Analysis analysis, Scope scope)
+    {
+        return new ScopeAware<T>(analysis, scope, node);
+    }
+
+    public T getNode()
+    {
+        return node;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return hash(node);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        ScopeAware<T> other = (ScopeAware<T>) o;
+        checkArgument(scope.isLocalScope(other.scope) || other.scope.isLocalScope(scope), "Expressions must be in the same local scope");
+
+        return equal(node, other.node);
+    }
+
+    @Override
+    public String toString()
+    {
+        return "ScopeAware(" + node + ")";
+    }
+
+    private boolean equal(Node left, Node right)
+    {
+        if (left instanceof Expression && right instanceof Expression) {
+            Expression leftExpression = (Expression) left;
+            Expression rightExpression = (Expression) right;
+            if (analysis.isColumnReference(leftExpression) && analysis.isColumnReference(rightExpression)) {
+                ResolvedField leftField = analysis.getResolvedField(leftExpression);
+                ResolvedField rightField = analysis.getResolvedField(rightExpression);
+
+                Scope leftScope = leftField.getScope();
+                Scope rightScope = rightField.getScope();
+
+                if (isLocalScope(leftScope) || isLocalScope(rightScope)) {
+                    // At least one of the fields comes from the same scope as the original expression, so compare by field id
+                    return leftField.getFieldId().equals(rightField.getFieldId());
+                }
+
+                // Otherwise, they are either:
+                // 1. Local to a subquery. In this case, the specific fields they reference don't matter as long as everything else is syntactically equal.
+                // 2. From a scope that is a parent to the base scope. In this case, the fields must necessarily be the same if they are syntactically equal.
+                return leftExpression.equals(rightExpression);
+            }
+        }
+
+        if (!left.shallowEquals(right)) {
+            return false;
+        }
+
+        List<? extends Node> leftChildren = left.getChildren();
+        List<? extends Node> rightChildren = right.getChildren();
+
+        if (leftChildren.size() != rightChildren.size()) {
+            return false;
+        }
+
+        for (int i = 0; i < leftChildren.size(); i++) {
+            if (!equal(leftChildren.get(i), rightChildren.get(i))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private int hash(Node node)
```

Comment:
it makes sense to precalculate the hash since it will be expensive to compute that hash every time